File: random_forest/javascript/rf.js
const NUM_TREES = 10;
const NUM_FEATURES = 5;

function loadData(filename) {
  const data = [];
  const reader = new FileReader(filename);
  reader.onload = () => {
    const lines = reader.result.split("\n");
    for (const line of lines) {
      const features = line.split(",");
      const label = parseInt(features[0]);
      const featureValues = new Float32Array(features.slice(1));
      data.push({ label, featureValues });
    }
  };
}

function createRandomForest() {
  const treeLabels = [];
  for (let i = 0; i < NUM_TREES; i++) {
    treeLabels.push(Math.random() < 0.5 ? 0 : 1);
  }
  return treeLabels;
}

function classifyData(data, treeLabels) {
  const predictions = [];
  for (const dp of data) {
    const correctLabel = dp.label;
    const prediction = 0;
    for (const treeLabel of treeLabels) {
      if (treeLabel === correctLabel) {
        prediction++;
      }
    }
    predictions.push(prediction > NUM_TREES / 2 ? 1 : 0);
  }
  return predictions;
}

function calculateAccuracy(predictions, data) {
  const correct = predictions.reduce((acc, p) => acc + (p === data[p].label ? 1 : 0), 0);
  return (correct / predictions.length) * 100;
}

function main() {
  const data = loadData("data.csv");
  const treeLabels = createRandomForest();
  const predictions = classifyData(data, treeLabels);
  const accuracy = calculateAccuracy(predictions, data);
  console.log("Accuracy: " + accuracy + "%");
}

main();

File: random_forest/scala/rf.scala
import scala.util.Random

val NUM_TREES = 10
val NUM_FEATURES = 5

def loadData(filename: String): Seq[(Int, Seq[Double])] = {
  val data = scala.io.Source.fromFile(filename).getLines().map { line =>
    val features = line.split(",").map(_.toDouble)
    (features.head.toInt, features.tail)
  }.toSeq
  data
}

def createRandomForest(): Seq[Int] = {
  val treeLabels = Seq.fill(NUM_TREES)(Random.nextInt(2))
  treeLabels
}

def classifyData(data: Seq[(Int, Seq[Double])], treeLabels: Seq[Int]): Seq[Int] = {
  val predictions = data.map { dp =>
    val correctLabel = dp._1
    val prediction = 0
    treeLabels.foreach { treeLabel =>
      if (treeLabel == correctLabel) {
        prediction += 1
      }
    }
    prediction > NUM_TREES / 2
  }
  predictions
}

def calculateAccuracy(predictions: Seq[Int], data: Seq[(Int, Seq[Double])]): Double = {
  val correct = predictions.zip(data).count { case (p, d) => p == d._1 }
  correct.toDouble / data.length.toDouble * 100
}

def main(args: Array[String]): Unit = {
  val data = loadData("data.csv")
  val treeLabels = createRandomForest()
  val predictions = classifyData(data, treeLabels)
  val accuracy = calculateAccuracy(predictions, data)
  println(s"Accuracy: ${accuracy}%")
}
